### PR TITLE
Feature/webengine projection mode

### LINKED ISSUE #234: [SDL 0273] WebEngine Projection mode
Proposal: [WebEngine Projection mode](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0273-webengine-projection-mode.md)

> This proposal is created based on discussion in https://github.com/smartdevicelink/sdl_evolution/issues/767.
It extends [SDl-0240 - WebEngine support for SDL JavaScript](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0240-sdl-js-pwa.md) to allow using the WebEngine app's web view for the HMI in addition to the other available SDL templates.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/912

Steering Committee Decision:

>The Steering Committee voted to accept this proposal with the revisions outlined in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/912#issuecomment-637749152).

> The proposal .md file was updated to reflect these revisions on 6/8/2020.

### PR DESCRIPTION
**Pull request by AKalinich-Luxoft**
Fixes #234
Implements [SDL 0273](https://github.com/smartdevicelink/sdl_javascript_suite/issues/234)

This PR is **not ready** for review.

### Testing Plan
Will be tested manually according to checklist

### Summary
- Update library for the new App HMI type WEB_VIEW
- Update minified libraries for the new App HMI type WEB_VIEW
- Update webengine/hello-sdl application to be ready for the new App HMI type WEB_VIEW

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: lib/js/src/manager/lifecycle/_LifecycleManager.js [hunk 1]
```diff
@@ -506,7 +506,7 @@ class _LifecycleManager {
     }
 }
 
-_LifecycleManager.MAX_RPC_VERSION = new Version(6, 0, 0);
+_LifecycleManager.MAX_RPC_VERSION = new Version(6, 2, 0);
```
Comment (SilvioDC-Ford) [inline lib/js/src/manager/lifecycle/_LifecycleManager.js:509]:
```suggestion
_LifecycleManager.MAX_RPC_VERSION = new Version(7, 0, 0);
```

File: examples/webengine/hello-sdl/manifest.js [hunk 2]
```diff
@@ -2,17 +2,17 @@
...
+            "appName": "Hallo WebEngine3",
             "appIcon": "./test_icon_1.png"
         }
     },
     "appVersion": "1.0.0",
-    "minRpcVersion": "6.0",
-    "minProtocolVersion": "5.0"
-};
\ No newline at end of file
+    "minRpcVersion": "6.2.0",
```
Comment (SilvioDC-Ford) [inline examples/webengine/hello-sdl/manifest.js:16]:
```suggestion
    "minRpcVersion": "7.0.0",
```

File: lib/js/src/rpc/enums/AppInterfaceUnregisteredReason.js [hunk 3]
```diff
@@ -143,6 +143,14 @@ class AppInterfaceUnregisteredReason extends Enum {
         return AppInterfaceUnregisteredReason._MAP.UNSUPPORTED_HMI_RESOURCE;
     }
 
+    /**
+     * Get the enum value for RESOURCE_CONSTRAINT.
+     * @returns {String} - The enum value.
+     */
```
Comment (SilvioDC-Ford) [inline lib/js/src/rpc/enums/AppInterfaceUnregisteredReason.js:149]: Please align with other libs. Take text of RPC spec.

File: lib/js/src/rpc/enums/AppHMIType.js [hunk 4]
```diff
@@ -143,6 +143,14 @@ class AppHMIType extends Enum {
         return AppHMIType._MAP.REMOTE_CONTROL;
     }
 
+    /**
+     * Get the enum value for WEB_VIEW.
+     * @returns {String} - The enum value.
+     */
```
Comment (SilvioDC-Ford) [inline lib/js/src/rpc/enums/AppHMIType.js:149]: Please align with other libs. Take text of RPC spec.

File: lib/js/src/rpc/enums/PredefinedLayout.js [hunk 5]
```diff
@@ -231,6 +231,16 @@ class PredefinedLayout extends Enum {
         return PredefinedLayout._MAP.LARGE_GRAPHIC_ONLY;
     }
 
+    /**
+     * Get the enum value for WEB_VIEW.
+     * Custom root template allowing in-vehicle WebEngine applications with appropriate permissions to show the
+     * application's own web view.
+     * @returns {String} - The enum value.
+     */
```
Comment (SilvioDC-Ford) [inline lib/js/src/rpc/enums/PredefinedLayout.js:239]: Please align with other libs. Take text of RPC spec.

File: lib/js/src/manager/lifecycle/_LifecycleManager.js (hunk 1 above)
Comment (AKalinich-Luxoft) [reply lib/js/src/manager/lifecycle/_LifecycleManager.js]: @SilvioDC-Ford fixed in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/8035899de3a2b59eae39bf9806b90bdabd46cdc7

File: examples/webengine/hello-sdl/manifest.js (hunk 2 above)
Comment (AKalinich-Luxoft) [reply examples/webengine/hello-sdl/manifest.js]: @SilvioDC-Ford fixed in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/8035899de3a2b59eae39bf9806b90bdabd46cdc7

File: lib/js/src/rpc/enums/AppInterfaceUnregisteredReason.js (hunk 3 above)
Comment (AKalinich-Luxoft) [reply lib/js/src/rpc/enums/AppInterfaceUnregisteredReason.js]: @SilvioDC-Ford fixed in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/07b2d177e818f95d584fb8f67e3638558f667bec

File: lib/js/src/rpc/enums/AppHMIType.js (hunk 4 above)
Comment (AKalinich-Luxoft) [reply lib/js/src/rpc/enums/AppHMIType.js]: @SilvioDC-Ford fixed in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/07b2d177e818f95d584fb8f67e3638558f667bec

File: lib/js/src/rpc/enums/PredefinedLayout.js (hunk 5 above)
Comment (AKalinich-Luxoft) [reply lib/js/src/rpc/enums/PredefinedLayout.js]: @SilvioDC-Ford fixed in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/07b2d177e818f95d584fb8f67e3638558f667bec

**Comment by AKalinich-Luxoft** [comment]
Develop branch has been merged into the feature branch to resolve conflicts in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/4c0c214e3409571448bcabe1227af92278d77bfd

**Comment by kshala-ford** [comment]
Thanks @AKalinich-Luxoft @SilvioDC-Ford 

Ford approves this PR

**Comment by SilvioDC-Ford** [comment]
@theresalech this PR is ready for review and is approved by Ford.

**Comment by iCollin** [comment]
@AKalinich-Luxoft would it be possible to add more buttons to this example app to support other interactions such as ScrollableMessage, Slider or PerformInteraction? These will be necessary for testing projection mode.

**Comment by AKalinich-Luxoft** [comment]
> @AKalinich-Luxoft would it be possible to add more buttons to this example app to support other interactions such as ScrollableMessage, Slider or PerformInteraction? These will be necessary for testing projection mode.

@iCollin please check for a new RPCs added for test application in https://github.com/smartdevicelink/sdl_javascript_suite/pull/264/commits/4512f262c13ee0c8e8bbbf45e035ec691afab910

**Comment by crokita** [comment]
While we continue our review of this PR, please pull in the new changes from the develop branch. We are removing the built files from the project so that merge conflicts will stop happening with them and the processing of contributing and reviewing code will be easier.
@AKalinich-Luxoft

**Comment by vladmu** [comment]
@crokita the feature branch is updated with new changes from the develop branch. Please review.

File: examples/webengine/hello-sdl/index.html [hunk 6]
```diff
@@ -119,86 +130,548 @@
...
+
+                    return rpcStruct; // leave it alone
+                }
+
+                const msg_type = SDL.rpc.enums.MessageType.keyForValue(message.getMessageType());
+                this._logMessage(prefix, `Rpc: "${message.getFunctionId()}" Type: "${msg_type}" Params: ${JSON.stringify(paramify(message))}`);
+            }
+
+            _logMessage(prefix, message, isError = false) {
+                const get_time = function() {
```
Comment (crokita) [inline examples/webengine/hello-sdl/index.html:259]: This isn't the only case where it happens, but please stick to the [Naming Convention Guidelines](https://github.com/smartdevicelink/sdl_javascript_suite/blob/develop/GUIDELINES.md#naming-convention): "Objects, functions, and instances should use camelCase."

File: examples/webengine/hello-sdl/index.html [hunk 7]
```diff
@@ -119,86 +130,548 @@
...
+
+            app.sendRpcRequest(get_create_choice_request())
+                .then( () => {
+                    app.sendRpcRequest(get_pi_request())
+                        .then( () => {
+                            setTimeout(() => {
+                                app.sendRpcRequest(get_del_choice_request());
+                            }, 10000);
+                        })
+                });
```
Comment (crokita) [inline examples/webengine/hello-sdl/index.html:538]: Use `await` syntax to make this snippet easier to read:
```js
            await app.sendRpcRequest(get_create_choice_request());
            await app.sendRpcRequest(get_pi_request())
            await new Promise(resolve => setTimeout(resolve, 10000));
            await app.sendRpcRequest(get_del_choice_request())
```

The event listener function is allowed to be an async function.

The functions being called should use camelCase.

File: lib/js/src/manager/LifecycleConfig.js [hunk 8]
```diff
@@ -366,15 +367,18 @@ class LifecycleConfig {
                     break;
                 case 'category':
                 case 'additionalCategories': {
+                    const appTypes = isCategoriesChanged ? this.getAppTypes() : [];
```
[the remark this answers is not in the document]
Comment (vladmu) [reply lib/js/src/manager/LifecycleConfig.js:370]: @crokita as far as we use only one `appTypes` property for setting app categories but two separate properties `category` and `additionalCategories` in the manifest, we need to control if the category has already come from one of those properties. So the second time the `appTypes` needs to be changed we get previously set `appTypes` to merge them with the new values. To be honest, the logic was not changed from the previous implementation except this `isCategoriesChanged` flag fixes the problem when the `DEFAULT` category was always added into the list even if we just set it as `WEB_VIEW` int the manifest.

**Comment by vladmu** [comment]
@crokita all issues were processed. Please review.

**Pull request merged**